File: src/sp/it/pl/gui/objects/grid/GridCell.java
package sp.it.pl.gui.objects.grid;

import javafx.scene.control.IndexedCell;
import javafx.scene.control.Skin;
import sp.it.pl.util.access.V;

/**
 * A cell of {@link GridView}. It contains single item in the {@link sp.it.pl.gui.objects.grid.GridView#getItemsShown()} list.
 *
 * @see GridView
 */
public class GridCell<T, F> extends IndexedCell<T> {

	/**
	 * {@link sp.it.pl.gui.objects.grid.GridView} this cell belongs to.
	 */
	public final V<GridView<T,F>> gridView = new V<>(null);

	public GridCell() {
		getStyleClass().add("grid-cell");
	}

	@Override
	public void updateIndex(int i) {
		if (getIndex()==i) return;
		super.updateIndex(i);
	}

	void update(T item, boolean isSelected) {
		updateItem(item, item==null);
		updateSelected(isSelected);
	}

	@Override
	protected Skin<?> createDefaultSkin() {
		return new GridCellSkin<>(this);
	}

}

File: src/player/main/sp/it/pl/ui/objects/grid/GridCell.kt
package sp.it.pl.ui.objects.grid

import javafx.scene.control.IndexedCell
import sp.it.util.access.V
import sp.it.util.ui.pseudoclass

/** A cell of [GridView], contains single item in the [sp.it.pl.ui.objects.grid.GridView.itemsShown] list. */
open class GridCell<T: Any, F: Any>: IndexedCell<T>() {

   /** [GridView] this cell belongs to. */
   val gridView = V<GridView<T, F>?>(null)

   init {
      styleClass += "grid-cell"
   }

   override fun updateIndex(i: Int) {
      if (index==i) return
      super.updateIndex(i)
   }

   fun update(i: Int, item: T?, isSelected: Boolean) {
      updateIndex(i)
      updateItem(item, item==null)
      updateSelected(isSelected)
   }

   override fun createDefaultSkin() = GridCellSkin(this)

   /** Dispose of this cell with the intention of never being used again. Called automatically in when grid skin disposes. */
   open fun dispose() {}

   companion object {
      @JvmField var pseudoclassFullWidth = pseudoclass("full-width")
   }
}